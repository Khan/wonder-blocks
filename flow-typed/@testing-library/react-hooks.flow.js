// @flow
/**
 * This file is based on:
 * - https://github.com/testing-library/react-hooks-testing-library/blob/main/src/types/index.ts
 * - https://github.com/testing-library/react-hooks-testing-library/blob/main/src/types/react.ts
 * - https://github.com/testing-library/react-hooks-testing-library/blob/main/src/pure.ts
 */
 declare module "@testing-library/react-hooks" {
    declare export type Renderer<TProps> = {|
        render: (props?: TProps) => void,
        rerender: (props?: TProps) => void,
        unmount: () => void,
        act: Act,
    |};

    declare export type ServerRenderer<TProps> = {|
        ...Renderer<TProps>,
        hydrate: () => void,
    |};

    declare export type RendererProps<TProps, TResult> = {|
        callback: (props: TProps) => TResult,
        setError: (error: Error) => void,
        setValue: (value: TResult) => void,
    |};

    declare export type CreateRenderer<
        TProps,
        TResult,
        TRendererOptions: {...},
        TRenderer: Renderer<TProps>,
    > = (
        props: RendererProps<TProps, TResult>,
        options: TRendererOptions,
    ) => TRenderer;

    declare export type RenderResult<TValue> = {|
        +all: Array<TValue | Error>,
        +current: TValue,
        +error?: Error,
    |};

    declare export type ResultContainer<TValue> = {|
        result: RenderResult<TValue>,
    |};

    declare export type WaitOptions = {|
        interval?: number | false,
        timeout?: number | false,
    |};

    declare export type WaitForOptions = WaitOptions;
    declare export type WaitForValueToChangeOptions = WaitOptions;
    declare export type WaitForNextUpdateOptions = WaitOptions["timeout"];

    declare export type WaitFor = (
        callback: () => boolean | void,
        options?: WaitForOptions,
    ) => Promise<void>;
    declare export type WaitForValueToChange = (
        selector: () => mixed,
        options?: WaitForValueToChangeOptions,
    ) => Promise<void>;
    declare export type WaitForNextUpdate = (
        options?: WaitForNextUpdateOptions,
    ) => Promise<void>;

    declare export type AsyncUtils = {|
        waitFor: WaitFor,
        waitForValueToChange: WaitForValueToChange,
        waitForNextUpdate: WaitForNextUpdate,
    |};

    declare export type RenderHookResult<
        TProps,
        TValue,
        TRenderer: Renderer<TProps> = Renderer<TProps>,
    > = {|
        ...ResultContainer<TValue>,
        // Omit<Renderer<TProps>, 'render' | 'act'>
        ...$Diff<
            Renderer<TProps>,
            {|
                render: Renderer<TProps>["render"],
                act: Renderer<TProps>["act"],
            |},
        >,
        // Omit<TRenderer, keyof Renderer<TProps>>
        ...$Diff<TRenderer, Renderer<TProps>>,
        ...AsyncUtils,
    |};

    declare export type ServerRenderHookResult<
        TProps,
        TValue,
        TRenderer: ServerRenderer<TProps> = ServerRenderer<TProps>,
    > = RenderHookResult<TProps, TValue, TRenderer>;

    declare export type Act = {|
        (callback: () => Promise<void>): Promise<void>,
        (callback: () => void): void,
    |};

    declare export type CleanupCallback = () => Promise<void> | void;

    declare export type WrapperComponent<TProps> = React$ComponentType<TProps>;

    declare export type RendererOptions<TProps> = {|
        wrapper?: WrapperComponent<TProps>,
    |};

    declare export type RenderHookOptions<TProps> = {|
        initialProps?: TProps,
        wrapper?: WrapperComponent<any>,
    |};

    declare export type ReactHooksRenderer = {|
        renderHook: <TProps, TResult>(
            callback: (props: TProps) => TResult,
            options?: RenderHookOptions<TProps>,
        ) => RenderHookResult<TProps, TResult>,
        act: Act,
        cleanup: () => Promise<void>,
        addCleanup: (callback: CleanupCallback) => () => void,
        removeCleanup: (callback: CleanupCallback) => void,
        suppressErrorOutput: () => () => void,
    |};

    declare export type ReactHooksServerRenderer = {|
        // Omit<ReactHooksRenderer, 'renderHook'>
        ...$Diff<
            ReactHooksRenderer,
            {|
                renderHook: ReactHooksRenderer["renderHook"],
            |},
        >,
        renderHook: <TProps, TResult>(
            callback: (props: TProps) => TResult,
            options?: RenderHookOptions<TProps>,
        ) => ServerRenderHookResult<TProps, TResult>,
    |};

    declare export function renderHook<TProps, TResult>(
        callback: (props: TProps) => TResult,
        options?: RenderHookOptions<TProps>,
    ): RenderHookResult<TProps, TResult>;
    declare export var act: Act;
    declare export function cleanup(): Promise<void>;
    declare export function addCleanup(callback: CleanupCallback): () => void;
    declare export function removeCleanup(callback: CleanupCallback): void;
    declare export function suppressErrorOutput(): () => void;
}
