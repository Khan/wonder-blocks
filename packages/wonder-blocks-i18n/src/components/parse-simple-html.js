// @flow
/**
 * Simple parser for a subset of HTML.
 *
 * This is an internal function intended to be used by I18nInlineMarkup.
 */

export type Tag = {|
    type: "tag",
    tag: string,
    children: string | null,
|};

export type Text = {|
    type: "text",
    text: string,
|};

export type SimpleHtmlNode = Tag | Text;

function parseSimpleHTML(html: string): $ReadOnlyArray<SimpleHtmlNode> {
    // This is a regex that can capture the following kinds of things:
    //  - self-closing tags: (e.g., <myimg />)
    //  - non self-closing tags. This regex captures an opening tag, and the
    //    next tag (regardless of whether it is opening, closing, or
    //    self-closing). We make sure, later in this function, that the second
    //    tag is a closing tag of the same type as the opening tag.
    //  - text (i.e., parts of the string that aren't tags).
    //
    // NOTE: Don't move simpleHtmlRegex out of this function. Regexes are
    // stateful when used with `exec`, so we need to recreate it every time
    // parseSimpleHTML is called.
    //
    // TODO(joshuan): This would be a great opportunity for named captures.
    // If those are ever implemented in browsers we care about, this should
    // be rewritten to use them.
    const simpleHtmlRegex =
        /(<([^>^/]+)\s*\/>)|(<([^>]+)>([^<]*)<(\/?)([^>]+)>)|([^<]+)/gm;
    // simpleHtmlRegex has numbered captures. The following are names for those
    // captures. If you ever need to modify the above regex and need to
    // figure out what the new capture numbers are, https://regex101.com
    // must be helpful.
    const SELF_CLOSE_NAME = 2;
    const OPEN_NAME = 4;
    const CHILDREN = 5;
    const CLOSING_SLASH = 6;
    const CLOSING_NAME = 7;
    const TEXT = 8;

    html = html.trim();

    const result = [];

    let match;
    while ((match = simpleHtmlRegex.exec(html))) {
        if (match[SELF_CLOSE_NAME] != null) {
            const tag = match[SELF_CLOSE_NAME].trim();
            if (tag.includes(" ")) {
                throw new Error(
                    "I18nInlineMarkup: expected a tag without " +
                        "attributes, but received: " +
                        `<${tag}/>`,
                );
            }
            result.push({
                type: "tag",
                tag,
                children: null,
            });
        } else if (match[OPEN_NAME] != null) {
            const tag = match[OPEN_NAME].trim();
            if (tag.includes(" ")) {
                throw new Error(
                    "I18nInlineMarkup: expected a tag without " +
                        "attributes, but received: " +
                        `<${match[OPEN_NAME]}>`,
                );
            }

            if (match[CLOSING_SLASH] !== "/") {
                throw new Error(
                    `I18nInlineMarkup: nested tags are not ` +
                        `supported, but <${match[CLOSING_NAME]}> is nested underneath ` +
                        `<${tag}>.`,
                );
            }

            const closingTag = match[CLOSING_NAME].trim();
            if (tag !== closingTag) {
                throw new Error(
                    `I18nInlineMarkup: expected closing tag ` +
                        `</${match[4]}>, but got </${match[7]}>`,
                );
            }
            result.push({
                type: "tag",
                tag,
                children: match[CHILDREN],
            });
        } else if (match[TEXT] != null) {
            result.push({
                type: "text",
                text: match[TEXT],
            });
        } else {
            throw new Error(
                "I18nInlineMarkup: unknown error (maybe you have an extra '<')?",
            );
        }
    }

    if (
        result.length === 1 &&
        // A tag is allowed to wrap all NL text in the html.
        (result[0].type === "text" || !result[0].children)
    ) {
        throw new Error("Unnecessary use of I18nInlineMarkup.");
    }
    return result;
}

export default parseSimpleHTML;
