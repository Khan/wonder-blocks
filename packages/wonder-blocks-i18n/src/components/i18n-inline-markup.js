// @flow

/**
 * Render some HTML, using functions passed in as props to render tags.
 *
 * ADR: https://docs.google.com/document/d/15EP6E_nxrEhgb9jlR3v7S93S22E3KAH_XMqN2gBqyUQ/edit
 *
 * Why?
 * ====
 *
 * Sometimes strings that you want to translate have links, words that are
 * bolded, etc.
 *
 * Historically, developers have chopped up these strings into bits without
 * HTML and stitched them back together manually. This occasionally worked.
 * Other times, a bug prevented the strings from being translated. When they
 * were translatable, sometimes we'd get translations with mismatched verb
 * tenses or other grammatical errors. Developers sometimes gave up and
 * resorted to using dangerouslySetInnerHTML, which isn't great either -- it
 * gives volunteers the ability to add XSS and it doesn't allow us to use our
 * fancy client-side links.
 *
 * With <I18nInlineMarkup>, things are better. You pass in a full translated
 * string, with HTML tags. For every kind of tag in the child string, you also
 * pass in a function which, given children, renders that kind of tag.
 *
 * For example, here's one way you could use <I18nInlineMarkup> to render a
 * disclaimer about our terms and conditions and privacy policy.
 *
 * ```
 * <I18nInlineMarkup
 *     link1={node => <a href="/terms">{node}</a>}
 *     link2={node => <a href="/privacy">{node}</a>}
 *     link3={node => <a href="/gdpr">{node}</a>}
 * >
 *     {i18n._(
          "By existing, you agree to our <link1>terms and conditions</link1>" +
              "and <link2>privacy policy</link2>. If you are a resident " +
              "of the European Union, you have <link3>options</link3>.",
 *     )}
 * </I18nInlineMarkup>
 * ```
 *
 * Use
 * ===
 *
 * <I18nInlineMarkup> accepts a single HTML-like string as children. This
 * string must not have any HTML attributes or nested tags. You must pass in
 * a render prop for every kind of tag in this string.
 *
 * Any numeric or named HTML codes will be escaped. That is, if children
 * contains `&mdash;` or `&#8212`, the text `&mdash;` or `&#8212;` will
 * show up to the end user. Use unicode escape sequences (e.g., \u2014)
 * instead.
 *
 * Here's a trivial use of I18nInlineMarkup:
 *
 * ```
 * <I18nInlineMarkup b={node => <b>{node}</b>}>
 *   {i18n._("Some <b>markup</b>")}
 * </I18nInlineMarkup>
 * ```
 *
 * Note that `b` doesn't need to actually return a `<b>` tag. It could, for
 * example, return a Wonder Blocks Label instead.
 *
 * You can the same kind of tag multiple times:
 *
 * ```
 * <I18nInlineMarkup b={node => <b>{node}</b>}>
 *   {i18n._("Some <b>markup</b> that I <b>wrote</b>")}
 * </I18nInlineMarkup>
 * ```
 *
 * Since attributes aren't supported, if you want to support different props
 * on the same kind of component, use multiple tag names. See the first
 * example in this comment.
 */

import * as React from "react";

import {parseSimpleHTML} from "./parse-simple-html.js";
import type {SimpleHtmlNode} from "./parse-simple-html.js";

type Props = {|
    /**
     * A translated string.
     *
     * TODO(joshuan): if we ever add a type for translated strings, replace
     * "string" with that type.
     */
    children: string,
    /**
     * A function which takes each top-level text or rendered tag,
     * and returns an element that wraps it.
     *
     * `type` is "text" if the element is text, and a pseudotag, like
     * "newline", or "cirlced-box" otherwise.
     *
     * We use this on the LOHP and marketing pages to provide a
     * background around text when they are on top of an illustration.
     *
     * i is the index of the text or tag.
     */
    elementWrapper?: (s: React.Node, type: string, i: number) => React.Node,

    onError?: (e: Error) => React.Node,

    [tag: string]: (content: string) => React.Node,
|};

export class I18nInlineMarkup extends React.PureComponent<Props> {
    /**
     * If an error occurs, we either call the onError prop, or throw the
     * error.
     */
    handleError(error: Error): React.Node {
        const {onError} = this.props;

        if (onError) {
            return onError(error);
        }

        throw error;
    }

    render(): React.Node {
        const {children, elementWrapper, ...renderers} = this.props;
        let tree: $ReadOnlyArray<SimpleHtmlNode>;
        try {
            tree = parseSimpleHTML(children);
        } catch (e) {
            return this.handleError(e);
        }
        const nodes: Array<React.Node> = tree.map((node, i) => {
            if (node.type === "text") {
                if (elementWrapper) {
                    return (
                        <React.Fragment key={i}>
                            {elementWrapper(node.text, "text", i)}
                        </React.Fragment>
                    );
                }
                return node.text;
            }
            if (node.type === "tag") {
                const renderer = renderers[node.tag];
                if (!renderer) {
                    return this.handleError(
                        new Error(
                            `I18nInlineMarkup: missing render prop for ${node.tag}`,
                        ),
                    );
                }
                if (elementWrapper) {
                    return (
                        <React.Fragment key={i}>
                            {elementWrapper(
                                /**
                                 * TODO(somewhatabstract, JIRA-XXXX):
                                 * node.children can be null but renderer does
                                 * not accept null.
                                 */
                                // $FlowFixMe[incompatible-call]
                                renderer(node.children),
                                node.tag,
                                i,
                            )}
                        </React.Fragment>
                    );
                }
                return (
                    <React.Fragment key={i}>
                        {
                            /**
                             * TODO(somewhatabstract, JIRA-XXXX):
                             * node.children can be null but renderer does
                             * not accept null.
                             */
                            // $FlowFixMe[incompatible-call]
                            renderer(node.children)
                        }
                    </React.Fragment>
                );
            }

            // istanbul ignore
            return this.handleError(new Error("Unknown child type."));
        });
        return nodes;
    }
}
