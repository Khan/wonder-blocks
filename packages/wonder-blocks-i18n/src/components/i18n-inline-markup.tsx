/**
 * Render some HTML, using functions passed in as props to render tags.
 *
 * ADR: https://docs.google.com/document/d/15EP6E_nxrEhgb9jlR3v7S93S22E3KAH_XMqN2gBqyUQ/edit
 *
 * Why?
 * ====
 *
 * Sometimes strings that you want to translate have links, words that are
 * bolded, etc.
 *
 * Historically, developers have chopped up these strings into bits without
 * HTML and stitched them back together manually. This occasionally worked.
 * Other times, a bug prevented the strings from being translated. When they
 * were translatable, sometimes we'd get translations with mismatched verb
 * tenses or other grammatical errors. Developers sometimes gave up and
 * resorted to using dangerouslySetInnerHTML, which isn't great either -- it
 * gives volunteers the ability to add XSS and it doesn't allow us to use our
 * fancy client-side links.
 *
 * With <I18nInlineMarkup>, things are better. You pass in a full translated
 * string, with HTML tags. For every kind of tag in the child string, you also
 * pass in a function which, given children, renders that kind of tag.
 *
 * For example, here's one way you could use <I18nInlineMarkup> to render a
 * disclaimer about our terms and conditions and privacy policy.
 *
 * ```
 * <I18nInlineMarkup
 *     link1={node => <a href="/terms">{node}</a>}
 *     link2={node => <a href="/privacy">{node}</a>}
 *     link3={node => <a href="/gdpr">{node}</a>}
 * >
 *     {i18n._(
          "By existing, you agree to our <link1>terms and conditions</link1>" +
              "and <link2>privacy policy</link2>. If you are a resident " +
              "of the European Union, you have <link3>options</link3>.",
 *     )}
 * </I18nInlineMarkup>
 * ```
 *
 * Use
 * ===
 *
 * <I18nInlineMarkup> accepts a single HTML-like string as children. This
 * string must not have any HTML attributes or nested tags. You must pass in
 * a render prop for every kind of tag in this string.
 *
 * Any numeric or named HTML codes will be escaped. That is, if children
 * contains `&mdash;` or `&#8212`, the text `&mdash;` or `&#8212;` will
 * show up to the end user. Use unicode escape sequences (e.g., \u2014)
 * instead.
 *
 * Here's a trivial use of I18nInlineMarkup:
 *
 * ```
 * <I18nInlineMarkup b={node => <b>{node}</b>}>
 *   {i18n._("Some <b>markup</b>")}
 * </I18nInlineMarkup>
 * ```
 *
 * Note that `b` doesn't need to actually return a `<b>` tag. It could, for
 * example, return a Wonder Blocks Label instead.
 *
 * You can the same kind of tag multiple times:
 *
 * ```
 * <I18nInlineMarkup b={node => <b>{node}</b>}>
 *   {i18n._("Some <b>markup</b> that I <b>wrote</b>")}
 * </I18nInlineMarkup>
 * ```
 *
 * Since attributes aren't supported, if you want to support different props
 * on the same kind of component, use multiple tag names. See the first
 * example in this comment.
 */

import * as React from "react";

import {parseSimpleHTML} from "./parse-simple-html";
import type {SimpleHtmlNode} from "./parse-simple-html";

type Props = {
    /**
     * A translated string.
     *
     * TODO(joshuan): if we ever add a type for translated strings, replace
     * "string" with that type.
     */
    // @ts-expect-error [FEI-5019] - TS2411 - Property 'children' of type 'string' is not assignable to 'string' index type '(content: string) => ReactElement<any, string | JSXElementConstructor<any>>'.
    children: string;
    /**
     * A function which takes each top-level text or rendered tag,
     * and returns an element that wraps it.
     *
     * `type` is "text" if the element is text, and a pseudotag, like
     * "newline", or "cirlced-box" otherwise.
     *
     * We use this on the LOHP and marketing pages to provide a
     * background around text when they are on top of an illustration.
     *
     * i is the index of the text or tag.
     */
    // @ts-expect-error [FEI-5019] - TS2411 - Property 'elementWrapper' of type '((s: ReactNode, type: string, i: number) => ReactElement<any, string | JSXElementConstructor<any>>) | undefined' is not assignable to 'string' index type '(content: string) => ReactElement<any, string | JSXElementConstructor<any>>'.
    elementWrapper?: (
        s: React.ReactNode,
        type: string,
        i: number,
    ) => React.ReactElement;
    // @ts-expect-error [FEI-5019] - TS2411 - Property 'onError' of type '((e: Error) => ReactElement<any, string | JSXElementConstructor<any>>) | undefined' is not assignable to 'string' index type '(content: string) => ReactElement<any, string | JSXElementConstructor<any>>'.
    onError?: (e: Error) => React.ReactElement;
    [tag: string]: (content: string) => React.ReactElement;
};

export class I18nInlineMarkup extends React.PureComponent<Props> {
    /**
     * If an error occurs, we either call the onError prop, or throw the
     * error.
     */
    handleError(error: Error): React.ReactElement {
        const {onError} = this.props;

        if (onError) {
            return onError(error);
        }

        throw error;
    }

    render(): React.ReactElement {
        const {children, elementWrapper, ...renderers} = this.props;
        let tree: ReadonlyArray<SimpleHtmlNode>;
        try {
            tree = parseSimpleHTML(children);
        } catch (e: any) {
            return this.handleError(e);
        }
        const nodes: Array<React.ReactNode> = tree.map((node, i) => {
            if (node.type === "text") {
                if (elementWrapper) {
                    return (
                        <React.Fragment key={i}>
                            {elementWrapper(node.text, "text", i)}
                        </React.Fragment>
                    );
                }
                return node.text;
            }
            if (node.type === "tag") {
                const renderer = renderers[node.tag];
                if (!renderer) {
                    return this.handleError(
                        new Error(
                            `I18nInlineMarkup: missing render prop for ${node.tag}`,
                        ),
                    );
                }
                if (elementWrapper) {
                    return (
                        <React.Fragment key={i}>
                            {elementWrapper(
                                /**
                                 * TODO(somewhatabstract, JIRA-XXXX):
                                 * node.children can be null but renderer does
                                 * not accept null.
                                 */
                                // @ts-expect-error [FEI-5019] - TS2345 - Argument of type 'string | null' is not assignable to parameter of type 'string'.
                                renderer(node.children),
                                node.tag,
                                i,
                            )}
                        </React.Fragment>
                    );
                }
                return (
                    <React.Fragment key={i}>
                        {
                            /**
                             * TODO(somewhatabstract, JIRA-XXXX):
                             * node.children can be null but renderer does
                             * not accept null.
                             */
                            // @ts-expect-error [FEI-5019] - TS2345 - Argument of type 'string | null' is not assignable to parameter of type 'string'.
                            renderer(node.children)
                        }
                    </React.Fragment>
                );
            }

            // istanbul ignore
            return this.handleError(new Error("Unknown child type."));
        });
        // @ts-expect-error [FEI-5019] - TS2322 - Type 'ReactNode[]' is not assignable to type 'ReactElement<any, string | JSXElementConstructor<any>>'.
        return nodes;
    }
}
