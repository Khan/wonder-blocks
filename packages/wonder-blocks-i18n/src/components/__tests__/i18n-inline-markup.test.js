// @flow
import * as React from "react";
import {render} from "@testing-library/react";

import * as ParseSimpleHTML from "../parse-simple-html";
import {I18nInlineMarkup} from "../i18n-inline-markup";
import {
    SingleShallowSubstitution,
    MultipleShallowSubstitution,
    ElementWrapper,
} from "../__docs__/i18n-inline-markup.stories";

describe("I18nInlineMarkup", () => {
    test("SingleShallowSubstitution", () => {
        const {container} = render(<SingleShallowSubstitution />);

        expect(container).toMatchSnapshot();
    });

    test("MultipleShallowSubstitution", () => {
        const {container} = render(<MultipleShallowSubstitution />);

        expect(container).toMatchSnapshot();
    });

    test("ElementWrapper", () => {
        const {container} = render(<ElementWrapper />);

        expect(container).toMatchSnapshot();
    });

    describe("errors", () => {
        beforeEach(() => {
            // React calls console.error() whenever there's an exception
            // thrown by a component.  We mock console.error() to avoid
            // error reports that we're expecting as part of these tests.
            jest.spyOn(console, "error").mockImplementation(() => {});
        });

        afterEach(() => {
            jest.clearAllMocks();
        });

        it("should throw an error if a render prop is missing", () => {
            const action = () =>
                render(
                    <I18nInlineMarkup>
                        {"Hello <b>world</b>!"}
                    </I18nInlineMarkup>,
                );

            expect(action).toThrowErrorMatchingInlineSnapshot(
                `"I18nInlineMarkup: missing render prop for b"`,
            );
        });

        it("should throw an error if `parseSimpleHTML()` throws", () => {
            jest.spyOn(ParseSimpleHTML, "parseSimpleHTML").mockImplementation(
                () => {
                    throw new Error("foo");
                },
            );

            const action = () =>
                render(
                    <I18nInlineMarkup>
                        {"Hello <b>world</b>!"}
                    </I18nInlineMarkup>,
                );

            expect(action).toThrowErrorMatchingInlineSnapshot(`"foo"`);
        });

        it("should call `onError` callback if set when an parse error is thrown", () => {
            jest.spyOn(ParseSimpleHTML, "parseSimpleHTML").mockImplementation(
                () => {
                    throw new Error("foo");
                },
            );
            const onErrorSpy = jest.fn().mockReturnValue("Hello world!");

            const {container} = render(
                <I18nInlineMarkup onError={onErrorSpy}>
                    {"Hello <b>world</b>!"}
                </I18nInlineMarkup>,
            );

            expect(onErrorSpy).toHaveBeenCalled();
            expect(container).toMatchSnapshot();
        });
    });
});
