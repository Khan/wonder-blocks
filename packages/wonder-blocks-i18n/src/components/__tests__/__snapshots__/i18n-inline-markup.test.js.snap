// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`I18nInlineMarkup ElementWrapper 1`] = `
<div>
  <span
    style="background: yellow;"
  >
    -6°C, 
  </span>
  <span
    style="background: yellow;"
  >
    <span
      style="background: red;"
    >
      __
      <u>
        Sunny
      </u>
      __
    </span>
  </span>
  <span
    style="background: yellow;"
  >
    , Fells 
  </span>
  <span
    style="background: yellow;"
  >
    <span
      style="background: lightblue;"
    >
      *
      <i
        style="font-style: italic;"
      >
        like
      </i>
      *
    </span>
  </span>
  <span
    style="background: yellow;"
  >
    : 
  </span>
  <span
    style="background: yellow;"
  >
    <span
      style="background: red;"
    >
      __
      <u>
        -12
      </u>
      __
    </span>
  </span>
  <span
    style="background: yellow;"
  >
    ,  Wind: VR 5 km/h
  </span>
</div>
`;

exports[`I18nInlineMarkup MultipleShallowSubstitution 1`] = `
<div>
  -6°C, 
  __
  <u>
    Sunny
  </u>
  __
  , Fells 
  <span
    style="background: lightblue;"
  >
    *
    <i
      style="font-style: italic;"
    >
      like
    </i>
    *
  </span>
  : 
  __
  <u>
    -12
  </u>
  __
  ,  Wind: VR 5 km/h
</div>
`;

exports[`I18nInlineMarkup SingleShallowSubstitution 1`] = `
<div>
  -6°C, Sunny, Fells like: 
  [Underline:
  <u>
    -12
  </u>
  ]
  , Wind: VR 5 km/h
</div>
`;

exports[`I18nInlineMarkup errors should call \`onError\` callback if set when an parse error is thrown 1`] = `
<div>
  Hello world!
</div>
`;

exports[`I18nInlineMarkup errors should call \`onError\` callback if set when there's invalid markup 1`] = `
<div>
  This HTML is broken 
  An error occurred!
  , but here is fine.
</div>
`;
